### ISSUE TITLE
scrape: Add more tests validating HTTP requests made on scrape (and their headers)

### ISSUE THREAD

**Issue by bwplotka** [prometheus/prometheus#13556]
### Proposal

This is to ensure bug like in https://github.com/prometheus/prometheus/issues/13401#issuecomment-1931903806 has less chance to be made in the next Prometheus versions.

Idea by @hacklschorsch

> Edit: Add something like https://pypi.org/project/httplint/ maybe?

**Comment by machine424** [comment]
I'm sure I ran `2.49.0` and prometheus had no problem scraping itself. Maybe the content negotiation implementation used by golang client is broken/doesn't follow the RFC? (accepts `q>1`?).

It can also be useful to have some e2e tests where we scrape real targets (not just `httptest`'s where a lot of stuff is mocked) of different prom clients...

**Comment by hacklschorsch** [comment]
> I'm sure I ran 2.49.0 and prometheus had no problem scraping itself. 

Yes, Prometheus seems to be rather lenient on what it accepts.
We also were still successfully scraping node-exporter, but not our own service's `/metrics` endpoint.

**Comment by machine424** [comment]
Yep, all targets based on the golang client should have a "broken" content negotiation, we'll probably need to fix [that](https://github.com/prometheus/common/blob/773d5664eb8d228d2c6ba235b1af12fbf7ee9020/internal/bitbucket.org/ww/goautoneg/autoneg.go#L84) or move to another library: https://github.com/elnormous/contenttype seems to better stick to the RFC and seems to be better tested.

(having a standard implementation will be even better https://github.com/golang/go/issues/19307)

**Comment by hacklschorsch** [comment]
Already 12 days ago the hotfix [was merged to the current "stable" NixOS-23.11 nixpkgs](https://github.com/NixOS/nixpkgs/pull/288817).  So all in all a good three months this was broken in NixOS stable.  Good things come to those who wait ^^

**Comment by bboreham** [comment]
Hello from the bug-scrub!  
We felt it would be good to validate Prometheus' outgoing http during CI.

**Comment by sammyqtran** [comment]
Hello,

I’d like to contribute by extending the existing scrape tests to validate additional headers and request properties to help prevent regressions.

I’m drafting a PR with these changes. Could you please review it and let me know if this aligns with the overall plan and expectations?

**Comment by loglapa** [comment]
I took a fresh look at this before starting another implementation. The RFC 9110 q-value coverage is already on `main` via #14852, and `TestTargetScraperScrapeOK` now also validates the Accept variants and scrape timeout. The remaining User-Agent / GET / Accept-Encoding checks are already implemented in #17492, which is still open with requested changes.

To avoid duplicating that work, is #17492 still the intended path to close this issue, or is there a specific remaining HTTP request property/e2e scope maintainers would like covered in a separate PR?

**Comment by Gambit-Checkmate** [comment]
Hi, I’d like to help expand scrape request tests. I can add focused coverage for the request method and headers without changing scrape behavior. Is there a particular header or request path the maintainers would prefer to prioritize?
